src/locale/keymap-util.c

```diff
@@ -780,3 +781,137 @@ int x11_convert_to_vconsole(Context *c) {
 
         return modified;
 }
+
+bool locale_gen_check_available(void) {
+        struct stat st;
+        if (stat("/usr/sbin/locale-gen", &st) < 0)
+                return false;
+        if (stat("/etc/locale.gen", &st) < 0)
+                return false;
+        return true;
+}
+
+int locale_gen_enable_locale(char *locale[_VARIABLE_LC_MAX]) {
+        _cleanup_fclose_ FILE *fr = NULL;
+        _cleanup_fclose_ FILE *fw = NULL;
+        _cleanup_free_ char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false;
+        bool first_line;
+        struct stat st;
+        int r;
+        const char *LOCALE_GEN_CONF_FNAME = "/etc/locale.gen";
+
+        if (isempty(locale[VARIABLE_LANG]))
+                return -EINVAL;
+
+        if (stat(LOCALE_GEN_CONF_FNAME, &st) < 0)
+                return -errno;
+
+        fr = fopen(LOCALE_GEN_CONF_FNAME, "re");
+        if (!fr)
+                return -errno;
+
+        r = fopen_temporary(LOCALE_GEN_CONF_FNAME, &fw, &temp_path);
+        if (r < 0)
+                return r;
+        (void) fchmod(fileno(fw), 0644);
+
+        if (!strstr(locale[VARIABLE_LANG], ".") || endswith_no_case(locale[VARIABLE_LANG], "UTF-8")) {
+                if (asprintf(&locale_entry, "%s UTF-8", locale[VARIABLE_LANG]) < 0)
+                        return -ENOMEM;
+        } else {
+                /* We do not process non-UTF-8 locale */
+                return -EINVAL;
+        }
+
+        first_line = true;
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                char *line_locale;
+
+                r = read_line(fr, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (locale_enabled) {
+                        /* Just complete writing the file if the new locale was already enabled */
+                        fprintf(fw, "%s%s", first_line? "" : "\n", line);
```

Comment:
i don't really grok why you output the newline before each line instead of after?

Reply:
The config file ends with a newline, and in case there was no commented locale to uncomment, we want to just append it at the end without carrying over the previous newline.
Otherwise after a few iterations we'd end up with a file that looks like
```
# AAAA

BBB

CCC
```
Since I can know whether I have actually reached the end of the file without prior calculations or reading it into a buffer, I prepend the newline so in case we reach the end we can either add a terminating newline to the file or the new locale and then a newline so the config file stays neat-ish.
Does that make sense to you? Do you prefer a different solution?
I added a comment explaining this briefly, meanwhile.